pkg/sentry/kernel/msgqueue/msgqueue.go

```diff
@@ -0,0 +1,275 @@
+// Copyright 2021 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package msgqueue implements System V message queues.
+package msgqueue
```

Comment:
A significant portion of this patch is identical to code from the `shm` package. This seems like an opportunity to share the code.

Consider defining an IPC object with the shared properties such as `ID`, `creator`, and `owner`, and have `shm.Shm` and `msgqueue.Queue` use this object. You can define the `ipcutil` helper functions directly on this object, and move more common functionality over such as allocating a new id and the lookup logic in `FindOrCreate`.

Reply:
I moved `pkg/sentry/ipcutil` to `pkg/sentry/kernel/ipc`, and added `ipc.Object`, which holds fields common to all IPC objects, and has permission/ownership functionality defined on it.